include/pybind11/pytypes.h

```diff
@@ -753,6 +763,39 @@ inline bool PyIterable_Check(PyObject *obj) {
     }
 }
 
+template <typename T>
+bool PyIterableT_Check(PyObject *obj) {
+    static_assert(
+        is_generic_type<T>::value || is_pyobject<T>::value,
+        "iterable_t can only be used with pyobjects and generic types "
+        "(py::class_<T>)");
+    PyObject *iter = PyObject_GetIter(obj);
+    if (iter) {
+        if (iter == obj) {
+            // If they are the same, then that's bad! For now, just throw a
+            // cast error.
+            Py_DECREF(iter);
+            throw cast_error(
+                "iterable_t<T> cannot be used with exhaustible iterables "
+                "(e.g., iterators, generators).");
+        }
+        bool good = true;
+        // Now that we know that the iterable `obj` will not be exhausted,
+        // let's check the contained types.
+        for (handle h : handle(iter)) {
```

Comment:
Working: C++11 is struggling with forward reference here. Should see how hard it is to just use pure CPython API, or shift this def. to after `py::handle` is declared.

Reply:
Done 19e8992e - I think